torch/distributed/distributed_c10d.py

```diff
@@ -190,6 +195,17 @@ def __new__(cls, name: str):
             value = name.lower()
         return value
 
+    @classmethod
+    def get_default_backend_for_device(cls, device: str):
```

Comment:
I am working on a PR that allows easier backend integration with custom backend https://github.com/pytorch/pytorch/pull/102336 so I am not sure if we would still need to change the c10d backend to set for the default backend for device. I assume if you register the custom backend, it should automatically override the backend config if you initialize the world_pg first?

I think maybe we can allow passing in the `device_type` to device_mesh be sth like `cuda:non-nccl` and pass this to the `init_process_group` call https://github.com/pytorch/pytorch/blob/7b47cd0a6c206ef1de1ad74881942352089ddc72/torch/distributed/distributed_c10d.py#L270

Reply:
Oh actually if we don't specify anything in `init_process_group`, it would not take the custom registered backend and get it initialized... it seems like we do need this map that maps from device_type to backend, cc @H-Huang does it make sense to have this reverse map for custom backends?

Reply:
> I assume if you register the custom backend, it should automatically override the backend config if you initialize the world_pg first?

> I think maybe we can allow passing in the device_type to device_mesh be sth like cuda:non-nccl and pass this to the init_process_group call
 
> Oh actually if we don't specify anything in init_process_group, it would not take the custom registered backend and get it initialized

> it seems like we do need this map that maps from device_type to backend,

As mentioned here, after registered a custom backend, there are two ways to initialize the backend (indluding custom backend):
1. User use `init_process_group` before use them, or passes backend to `init_process_group`.
2. Maintain a map from device_type to backend, which is updated when registering the custom backend. When we don't specify anything in `init_process_group` , it determines whether each device and backend in the map is available, and then uses the available device-backend pairs to initialize the processgroup.

Reply:
By the way, a unified function can be used to determine whether the backend is available, such as `is_backend_available`, instead of a special case for each backend. `is_backend_available` supports the unified judgment of built-in and third-party registered backends. I have an implementation https://github.com/pytorch/pytorch/pull/101945